plugin_runner/plugin_runner.py

```diff
@@ -182,6 +189,38 @@ async def ReloadPlugins(
             yield ReloadPluginsResponse(success=True)
 
 
+def validate_effects(effects: list[Effect]) -> list[Effect]:
+    """Validates the effects based on predefined rules."""
+    autocomplete_effects = [
+        effect for effect in effects if effect.type == EffectType.AUTOCOMPLETE_SEARCH_RESULTS
+    ]
+
+    if len(autocomplete_effects) > 1:
+        log.warning("Only one AUTOCOMPLETE_SEARCH_RESULTS effect is allowed.")
+        return []
+
+    return effects
+
+
+def apply_effects_to_context(effects: list[Effect], event: Event) -> Event:
+    """Applies AUTOCOMPLETE_SEARCH_RESULTS effects to the event context.
+    If we are dealing with a search event, we need to update the context with the search results.
+    """
+    event_name = event.name
+
+    # Skip if the event is not a search event
+    if not event_name.endswith("__PRE_SEARCH") and not event_name.endswith("__POST_SEARCH"):
+        return event
+
+    for effect in effects:
+        if effect.type == EffectType.AUTOCOMPLETE_SEARCH_RESULTS:
+            event.context["results"] = json.loads(effect.payload)
+            # Stop processing effects if we've found a AUTOCOMPLETE_SEARCH_RESULTS
+            break
```

Comment:
I want to avoid iterating through the entire list; if an AUTOCOMPLETE_SEARCH_RESULTS effect is found, I want to exit the loop.

Reply:
I must have misunderstood the rule (see my other comment).

I was thinking that it would be illegal to return more than one search-related effect, but the way it looks, we're saying that the rule is:

* Only one search-related effect can be returned.
* If a search-related effect is returned, no other effects can be returned.

Is that the intent? If so, I think that's a good thing, since as I said before it's much easier to be more strict up front and relax things later.